### PR TITLE
Dev logger lambda

### PR DESCRIPTION
**Pull request by antonKrouglov**
I am writing a net.core worker service project. There is no Startup class there. Updating Log after host has been built does not work as expected as only child scopes are affected. I am configuring serilog in code using custom config file which stores things connection strings for SQL Server sink. The only option I have found is to hide initialization inside lamba. This is a real autofac-style,

Sample usage:

```
using Autofac;
using AutofacSerilogIntegration;
using Microsoft.Extensions.Options;
using Serilog.Core;

public class SerilogConfiguratorFromSettings : IStartable {

    private readonly IOptionsMonitor<MyLoggingSettingsFromConfig> _MyLoggingSettingsFromConfigLive;
    public Logger logger { get; private set; }

    public SerilogConfiguratorFromSettings(IOptionsMonitor<MyLoggingSettingsFromConfig> MyLoggingSettingsFromConfigLive)
        => _MyLoggingSettingsFromConfigLive = MyLoggingSettingsFromConfigLive;

    public void Start()
        => logger = SerilogConfigurator.GetLoggerFromSettings(_MyLoggingSettingsFromConfigLive.CurrentValue);

}

public class LoggingModule : Module {

    protected override void Load(ContainerBuilder builder) {
        builder
            .RegisterType<SerilogConfiguratorFromSettings>()
            .As<IStartable>()
            .SingleInstance() //config is read once
            //.InstancePerLifetimeScope() //this allows to reread config for each new lifetime scope
            ;

        builder.RegisterLogger(
            loggerLambda: c => c.Resolve<SerilogConfiguratorFromSettings>().logger
            , autowireProperties: true);
    }
}

```

### CODE REVIEW

**Comment by srogovtsev** [comment]
Can you give an example of the intended usage? 

In my experience, as you still have a place where you configure the container, you can configure the logger there.

**Comment by antonKrouglov** [comment]
@srogovtsev - example added. I need to compose Logger _before_ autofac container is build. Logger depends on IOptions<MyLoggingSettingsFromConfig> which are not available directly. In the [absence of Startup](https://github.com/dotnet/runtime/issues/36000) lambda is the only way to solve this chicken-egg problem

**Comment by srogovtsev** [comment]
I'd say this is quite a strange scenario. If I remember properly, MS recommends having `IConfiguration` available during service configuration. Worker services [seem](https://medium.com/@nickfane/introduction-to-worker-services-in-net-core-3-0-4bb3fc631225) to follow the usual pattern with the `IHostBuilder`, so you'll have access to `IConfiguration` in the container build stage. Why does it have to be in the module?

(frankly, from my experience, registering logger in the module can have strange effects on the system)

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs [hunk 1]
```diff
@@ -46,7 +50,7 @@ protected override void Load(ContainerBuilder builder)
             {
                 builder.Register(c =>
                 {
-                    LoggerProvider provider = new LoggerProvider(_logger);
+                    LoggerProvider provider = new LoggerProvider(_loggerLambda(c));
```
Comment (srogovtsev) [inline src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: You have to understand that this weill make the logger effectively singleton, so no configuration reloads to which you allude in your example.

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs [hunk 2]
```diff
@@ -76,9 +80,9 @@ protected override void Load(ContainerBuilder builder)
                         .FirstOrDefault(np => np.Name == TargetTypeParameterName && np.Value is Type);
 
                     if (targetType != null)
-                        return (_logger ?? Log.Logger).ForContext((Type)targetType.Value);
+                        return (_loggerLambda(c) ?? Log.Logger).ForContext((Type)targetType.Value);
 
-                    return _logger ?? Log.Logger;
+                    return _loggerLambda(c) ?? Log.Logger;
```
Comment (srogovtsev) [inline src/AutofacSerilogIntegration/ContextualLoggingModule.cs:85]: Here you're creating a logger. But who and where will dispose it?

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs (hunk 1 above)
Comment (antonKrouglov) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: Indeed. In this scenario I would get rid of LoggerProvider altogether as it solves the same goal as the lamda does. But this would be a too bold change to propose:)

Comment (srogovtsev) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: Even if you get rid of the logger provider (that, by itself, is not a big issue), what will you do with the lifetime?

**Comment by antonKrouglov** [comment]
>>MS recommends having IConfiguration available during service configuration
This would a bit oversimplified scenario. In my case it would a bad design to access IConfiguration directly. IConfiguration should be accessed by special helper classes and abstracted properly for the purposes of uni testing. In may case parts of the configuration are being pulled from external sources. Helper classes may throw. Helper should log to special basic bootstrap logger.

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs (hunk 2 above)
Comment (antonKrouglov) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:85]: Normally, I would expect autofac to take care of this when disposing either root scope or nested lifetime scope. This kind of simplifies things. But you are using .ExternallyOwned() for the reasons I do not understand.

**Comment by srogovtsev** [comment]
> > MS recommends having IConfiguration available during service configuration
> 
> This would a bit oversimplified scenario. 

That's the one that works for asp.net Core.

> In my case it would a bad design to access IConfiguration directly. IConfiguration should be accessed by special helper classes and abstracted properly for the purposes of uni testing. 

You can easily do that during service configuration. On the other hand, though, there is problem in unit-testing code relying on `IConfiguration`, because you can always inject in-memory configuration.

> In may case parts of the configuration are being pulled from external sources. 

This is what `IConfigurationSource` is for.

But all of this is a bit outside of the scope, I agree. The problem is, having `Startup` class wouldn't help you either: the startup cannot affect the container, it is already built. So if you depend on something from the container to build your logger, the only option would be resolving it, and that will in turn lead to interesting lifetime management issues.

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs (hunk 2 above)
Comment (srogovtsev) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:85]: (Not me, @nblumhardt )

The reasons are, I believe, quite simple. We're in this branch because `dispose` is `false`, which means the developer who passed the logger will be responsible for its disposal. Because it is only one instance, it is quite easy to achieve. But you change that to returning an unknown - possible new on every call! - logger instance from lambda, and that poses a question of disposal.

Please note that in your specific example the logger will not be (explicitly) disposed at all.

**Comment by srogovtsev** [comment]
Actually, I would agree that mixing dynamically provided loggers with provided instances leads to confusion. I suggest we forget about the current changes for a moment. What syntax and behavior, especially regarding lifetime management, would you use for dynamically provided loggers?

**Comment by antonKrouglov** [comment]
>The problem is, having Startup class wouldn't help you either
As per [this question](https://web.archive.org/web/20190711180322/https://stackoverflow.com/questions/55551654/serilog-mssqlserver-email-with-custom-subject-sinks) - it does help. I have tried to emulate this approach with autofac with auto-start component - no luck. Making any changes to Serilog.Log.Logger affect child scopes only. I have event tried [this](https://github.com/serilog/serilog/issues/485#issuecomment-153751157) - no luck. This may be related to how Serilog integrates with IHostBuilder and IWebHostBuilder.

**Comment by antonKrouglov** [comment]
>What syntax and behavior, especially regarding lifetime management, would you use for dynamically provided loggers?

I would expect that provider (DI container or something similar to `LoggerProvider`) cares about lifetime of dynamically provided loggers. However I do not have any deep understanding what happens beneath the Serilog/MEF surface when `ILogger<MyClass>` are being instantiated and injected to each consumer.

**Comment by srogovtsev** [comment]
> > The problem is, having Startup class wouldn't help you either
> 
> As per [this question](https://web.archive.org/web/20190711180322/https://stackoverflow.com/questions/55551654/serilog-mssqlserver-email-with-custom-subject-sinks) - it does help. 

Erm.

```cs
public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory, IOptions<EmailLoggerSettings> emailLoggerSettings, IOptions<SqlLoggerSettings> sqlLoggerSettings)
{
  Log.Logger = new LoggerConfiguration()
    //...
    .CreateLogger();

  loggerFactory.AddSerilog(); //this is important
}
```

The place that I marked as important - this is [`Microsoft.Extensions.Logging.ILoggerFactory`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.iloggerfactory?view=dotnet-plat-ext-3.1). You won't be able to resolve `Serilog.ILogger` from the service provider.

(nor will it affect the loggers for the services that are created by this time)

**Comment by srogovtsev** [comment]
> I would expect that provider (DI container or something similar to LoggerProvider) cares about lifetime of dynamically provided loggers. 

It can "care" only as much as you explain it to. There's no way the DI container can tell whether you want the the loggers to be alive for just one resolve or for a scope lifetime, or for container lifetime. Be that a regular service, that'd be easy - we have all those handy `SingleInstance`, `InstancePerLifetimeScope` and so forth, but how do you propose to do that with the lambda you have provided?

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs (hunk 1 above)
Comment (antonKrouglov) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: This is a good point. In the example above `SerilogConfiguratorFromSettings` should implement IDisposable and take care or logger field lifetime. Probably alternate approach would be: use LoggerConfiguration instance or lambda as it seems to be a lightweight object so GC can take care of it if needed.

**Comment by antonKrouglov** [comment]
Well, lambda may cause troubles. How about just changing defaults? Say when Logger is not provided (1) try to resolve LoggerConfiguration from container (2) and only then default to static instance? That would solve my problem as well.

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs (hunk 1 above)
Comment (srogovtsev) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: > In the example above `SerilogConfiguratorFromSettings` should implement IDisposable and take care or logger field lifetime.

This will take care of the disposal itself, but (a) `SerilogConfiguratorFromSettings` cannot have any `ILogger`-consuming dependencies and (b) you still get only single-instance behaviour. Or, more specifically, you _can_ try implementing per-scope, but the state of your loggers will be very unpredictable.

**Comment by srogovtsev** [comment]
> try to resolve LoggerConfiguration from container

What should be the lifetime of the `Logger` created from this instance?

**Comment by antonKrouglov** [comment]
> What should be the lifetime of the Logger created from this instance?

I would say .InstancePerLifetimeScope. This would force to reload settings on child scopes. If app extensively creates scopes this may lead to problems, Therefore there maybe makes sense to add a flag that switches between InstancePerLifetimeScope or SingleInstance. Ideally new logger should be re-created only when LoggerConfiguration change.

**Comment by srogovtsev** [comment]
> > What should be the lifetime of the Logger created from this instance?
> 
> I would say .InstancePerLifetimeScope. 

And as soon as you have at least one `SingleInstance` _consumer_ you effectively have a logger that never changes whatever you do.

Which is worse, as soon as you have to concurrent scopes (which is usual for a web application which creates a scope for every incoming request), you'll have two concurrent loggers. And let's say your logging configuration has a [MS SQL sink](https://github.com/serilog/serilog-sinks-mssqlserver), which is, for the purposes of this example, a [periodic batching sink](https://github.com/serilog/serilog-sinks-periodicbatching), which [creates a timer](https://github.com/serilog/serilog-sinks-periodicbatching/blob/40b743aea8da6d9c85059c63be13be642e0da5bc/src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs#L128) for every instance of the sink. So, two concurrent scopes - two timers, which will not ever fire, because the usual lifetime of the request is shorter the default two seconds interval. But at least SQL handles concurrent writes decently. If you configure your logger to write to a file, you'll have to specify concurrent access even if your application is the only one that uses the file.

> Ideally new logger should be re-created only when LoggerConfiguration change.

There's no easy way to do that in Autofac. Basically, you'll have to to it yourself, which is not an easy thing if you consider concurrency issues. And you'll still have the problem that the loggers that were created from previous configuration might be (and most probably are) still in use, because there's almost no way of affecting them.

What is funny, though, is that you can expand on the recipe from one of your links above, and create a proxying Logger which does nothing - no enrichment, no level filtering, nothing - and forwards every event to a custom sink which in turn sends them to an `ILogger` that you have wherever you want (probably in some static) and change in whichever way you want. The only downside of this solution is that the consumers will always assume that all logging levels are enabled.

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs (hunk 1 above)
Comment (antonKrouglov) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: > SerilogConfiguratorFromSettings cannot have any ILogger-consuming dependencies 

Indeed, here it is not possible. In scenario where you have more control over DI (wcf project for instance) - it can done.,
log4net sample:
```

using System;
using Autofac;
using JetBrains.Annotations;
using log4net;


/// <summary>
/// Called from either global.asax or AutofacServiceHostFactory
///     //configsProvider is similar to IConfigurationSource in .net core
///     var configsProvider = GetRegularConfigsProvider(); 
///     var bootstrap = new Bootstrap();
///     bootstrap.Run(configsProvider);
/// </summary>
public class Bootstrap {

    #region properties injected by DI

    /// <summary> Logger; can be null </summary>
    [CanBeNull] private ILog _log;

    /// <summary> Log is set by IoC Container </summary>
    [CanBeNull]
    public ILog Log {
        get => _log;
        set => _log = value;
    }

    /// <summary> ссылка на родительский контейнер autofac </summary>
    public ILifetimeScope Container { get; set; }

    #region конфирураторы настроек и соотв настройки - injected by autofac (one-by-one)

    public Log4netConfigurator ConfigLog4net { get; set; }

    public ServiceSettingsConfigurator ConfigServiceSettings { get; set; }

    #endregion конфирураторы настроек и соотв настройки - injected by autofac (one-by-one)

    public AutofacConfiguration ConfigAutofac { get; set; }

    #endregion properties injected by DI

    public void Run(ConfigsProvider configsProvider) {
        try {
            #region build initial autofac container

            //build initial autofac container and start chicken(egg) nested resolution
            if (Container == null) {
                _log?.Info("building initial autofac container...");
                var builder = new ContainerBuilder();
                builder.RegisterType<Bootstrap>().AsSelf().PropertiesAutowired();
                builder.RegisterInstance(configsProvider).As<ConfigsProvider>();
                Container = builder.Build();
                var nextBootstrap = Container.Resolve<Bootstrap>();
                if (nextBootstrap == null) throw new NullReferenceException(nameof(nextBootstrap));
                nextBootstrap.Run(configsProvider);
                return;
            }

            #endregion


            #region log4net

            //log4net
            if (ConfigLog4net == null) {
                _log?.Info("log4net init...");
                var scope = Container.BeginLifetimeScope(s => {
                    s.RegisterType<Log4netConfigurator>().AsSelf().SingleInstance();
                    s.RegisterModule<Log4netModule>();
                });
                ConfigLog4net = scope.Resolve<Log4netConfigurator>(); //force XmlConfigurator.Configure(); on the instance creation
                if (ConfigLog4net == null) throw new NullReferenceException(nameof(ConfigLog4net));
                var nextBootstrap = scope.Resolve<Bootstrap>();
                if (nextBootstrap == null) throw new NullReferenceException(nameof(nextBootstrap));
                nextBootstrap.Run(configsProvider);
                return;
            }

            if (_log == null) throw new NullReferenceException(nameof(_log));

            #endregion log4net


            #region ServiceSettings

            //ServiceSettings
            if (ConfigServiceSettings == null) {
                _log?.Info("creating ServiceSettingsConfigurator...");
                var scope = Container.BeginLifetimeScope(s => {
                    s.RegisterType<ServiceSettingsConfigurator>().AsSelf().SingleInstance();
                    s.RegisterModule<Log4netModule>(); //Log4netModule have to be registered on each inner scope
                });
                ConfigServiceSettings = scope.Resolve<ServiceSettingsConfigurator>();
                if (ConfigServiceSettings == null) throw new NullReferenceException(nameof(ConfigServiceSettings));
                var nextBootstrap = scope.Resolve<Bootstrap>();
                if (nextBootstrap == null) throw new NullReferenceException(nameof(nextBootstrap));
                nextBootstrap.Run(configsProvider);
                return;
            }

            #endregion ServiceSettings


            #region config autofac globally

            //config autofac globally
            if (ConfigAutofac == null) {
                _log?.Info("configuring autofac...");
                var scope = Container.BeginLifetimeScope(s => {
                    s.RegisterType<AutofacConfiguration>().AsSelf().PropertiesAutowired();
                    s.RegisterModule<Log4netModule>(); //Log4netModule have to be registered on each inner scope
                });
                ConfigAutofac = scope.Resolve<AutofacConfiguration>();
                if (ConfigAutofac == null) throw new NullReferenceException(nameof(ConfigAutofac));
                ConfigAutofac.Configure();
            }

            #endregion config autofac globally
        }
        catch (Exception ex) {
            _log?.Error("init error", ex);
            throw;
        }
    }
}
```

Comment (srogovtsev) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: Unfortunately, this example is too wordy for me to understand the particular points you're trying to make of it.

In `Microsoft.Extensions.Hosting`-based projects (and .net Core Worker service seem to be one of those) there's quite a lot of container control. From [the looks of it](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-3.1&tabs=visual-studio&WT.mc_id=-blog-shchowd#worker-service-template) you can do whatever you want as long as you return an `IHostBuilder` that can build an `IHost`.

**Comment by antonKrouglov** [comment]
OK, it makes sense in a broad context. In a narrow context, in my particular case I do not have any child scopes. If required, I may re-engeneer this Windows service to have them a few with a low concurrency. Log setting that may change during runtime in this narrow case is either a particular sink is on or off. Can this be achieved somehow?

**Comment by srogovtsev** [comment]
> Can this be achieved somehow?

Yes, the easiest to do this would be to use a configure everything statically as usual, and then use [LoggingLevelSwitch](https://nblumhardt.com/2014/10/dynamically-changing-the-serilog-level/).

File: src/AutofacSerilogIntegration/ContextualLoggingModule.cs (hunk 1 above)
Comment (antonKrouglov) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: The point is that you can build a container incrementally, onion-way. The key is to use .PropertiesAutowired() and recursive calls.

Comment (srogovtsev) [reply src/AutofacSerilogIntegration/ContextualLoggingModule.cs:53]: There's nothing that stops you from doing the same in your own `IHostBuilder` or `IHostedService`.

**Comment by antonKrouglov** [comment]
Thank you
